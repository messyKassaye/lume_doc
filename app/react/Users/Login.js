import PropTypes from 'prop-types';
import React from 'react';
import { browserHistory } from 'react-router';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import { Field, LocalForm, actions as formActions } from 'react-redux-form';

import { Icon } from 'UI';

import { t } from 'app/I18N';
import { isClient } from 'app/utils';
import { reconnectSocket } from 'app/socket';
import RouteHandler from 'app/App/RouteHandler';
import { reloadThesauri } from 'app/Thesauri/actions/thesaurisActions';

import auth from 'app/Auth';

const reloadHome = () => {
  window.location.assign('/');
};

export class Login extends RouteHandler {
  constructor(props, context) {
    super(props, context);
    this.state = { error: false, error2fa: false, recoverPassword: false, tokenRequired: false };
    this.submit = this.submit.bind(this);
    this.setLogin = this.setLogin.bind(this);
    this.setRecoverPassword = this.setRecoverPassword.bind(this);
  }

  attachDispatch(dispatch) {
    this.formDispatch = dispatch;
  }

  submit(credentials) {
    if (this.state.recoverPassword) {
      return this.recoverPassword(credentials.username);
    }

    return this.login(credentials);
  }

  recoverPassword(email) {
    this.setState({ recoverPassword: false, error: false });
    this.formDispatch(formActions.reset('loginForm'));
    return this.props.recoverPassword(email);
  }

  resolveSuccessfulLogin() {
    if (this.props.private) {
      reloadHome();
      return;
    }
    reconnectSocket();
    this.props.reloadThesauris();
    browserHistory.push('/');
  }

  async login(credentials) {
    try {
      await this.props.login(credentials);
      this.resolveSuccessfulLogin();
    } catch (err) {
      if (!this.state.tokenRequired && err.status === 409) {
        this.setState({ tokenRequired: true });
      } else {
        const { tokenRequired } = this.state;
        this.formDispatch(formActions.change('loginForm.token', undefined));
        const error2fa = tokenRequired;
        this.setState({ error: true, tokenRequired, error2fa });
      }
    }
  }

  setRecoverPassword() {
    this.formDispatch(formActions.reset('loginForm'));
    this.setState({ recoverPassword: true, error: false });
  }

  setLogin() {
    this.formDispatch(formActions.reset('loginForm'));
    this.setState({ recoverPassword: false, tokenRequired: false, error: false, error2fa: false });
  }

  render() {
    let submitLabel = this.state.recoverPassword
      ? t('System', 'Send recovery email')
      : t('System', 'Login button', 'Login');

    if (this.state.tokenRequired) {
      submitLabel = t('System', 'Verify');
    }

    return (
      <div className="content login-content">
        <div className="row">
          <div className="col-xs-12 col-sm-4 col-sm-offset-4" suppressHydrationWarning>
            <h1 className="login-title">
              <img src="/public/logo.svg" title="uwazi" alt="uwazi" />
            </h1>
            {isClient && (
              <LocalForm
                onSubmit={this.submit}
                model="loginForm"
                getDispatch={dispatch => {
                  this.formDispatch = dispatch;
                }}
              >
                {!this.state.tokenRequired && (
                  <>
                    <div
                      className={`form-group login-email${this.state.error ? ' has-error' : ''}`}
                    >
                      <Field model=".username">
                        <label className="form-group-label" htmlFor="username">
                          {this.state.recoverPassword ? t('System', 'Email') : t('System', 'User')}
                        </label>
                        <input type="text" name="username" id="username" className="form-control" />
                      </Field>
                    </div>
                    <div
                      className={`form-group login-password ${this.state.error ? 'has-error' : ''}${
                        this.state.recoverPassword ? ' is-hidden' : ''
                      }`}
                    >
                      <label className="form-group-label" htmlFor="password">
                        {t('System', 'Password')}
                      </label>
                      <Field model=".password">
                        <input
                          type="password"
                          name="password"
                          id="password"
                          className="form-control"
                        />
                      </Field>
                      <div className="form-text">
                        {this.state.error && <span>{t('System', 'Login failed')} - </span>}
                        <span
                          title={t('System', 'Forgot Password?', null, false)}
                          onClick={this.setRecoverPassword}
                          className={`button forgot-password ${
                            this.state.error ? 'label-danger' : ''
                          }`}
                        >
                          {t('System', 'Forgot Password?')}
                        </span>
                      </div>
                    </div>
                  </>
                )}
                {this.state.tokenRequired && (
                  <div
                    className={`form-group login-token${this.state.error2fa ? ' has-error' : ''}`}
                  >
                    <h5>{t('System', 'Two-step verification')}</h5>
                    <Field model=".token">
                      <label className="form-group-label" htmlFor="token">
                        {t('System', 'Authentication code')}
                      </label>
                      <input type="text" name="token" id="token" className="form-control" />
                      <div className="form-text">
                        <p>
                          {this.state.error2fa && (
                            <>
                              <Icon icon="exclamation-triangle" />
                              <span>{t('System', 'Two-factor verification failed')}</span>
                            </>
                          )}
                        </p>
                        <p>
                          Open the two-factor Authenticator app on your device <br />
                          to view your authentication code and verify your identity.
                        </p>
                        <p>
                          <span
                            onClick={this.setLogin}
                            className={`button forgot-password ${
                              this.state.error2fa ? 'label-danger' : ''
                            }`}
                          >
                            {t('System', 'Return to login')}
                          </span>
                        </p>
                      </div>
                    </Field>
                  </div>
                )}
                <p>
                  <button
                    type="submit"
                    className={`btn btn-block btn-lg ${
                      this.state.recoverPassword ? 'btn-success' : 'btn-primary'
                    }`}
                  >
                    {submitLabel}
                  </button>
                </p>
                {this.state.recoverPassword && (
                  <div className="form-text">
                    <span
                      title={t('System', 'Cancel', null, false)}
                      onClick={this.setLogin}
                      className="button cancel"
                    >
                      {t('System', 'Cancel')}
                    </span>
                  </div>
                )}
              </LocalForm>
            )}
          </div>
        </div>
      </div>
    );
  }
}

Login.propTypes = {
  login: PropTypes.func,
  recoverPassword: PropTypes.func,
  reloadThesauris: PropTypes.func,
};

export function mapStateToProps({ settings }) {
  return {
    private: settings.collection.get('private'),
  };
}

function mapDispatchToProps(dispatch) {
  return bindActionCreators(
    {
      login: auth.actions.login,
      recoverPassword: auth.actions.recoverPassword,
      reloadThesauris: reloadThesauri,
    },
    dispatch
  );
}

export default connect(mapStateToProps, mapDispatchToProps)(Login);
