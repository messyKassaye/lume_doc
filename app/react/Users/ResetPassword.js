import PropTypes from 'prop-types';
import React from 'react';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import { browserHistory } from 'react-router';
import RouteHandler from 'app/App/RouteHandler';
import queryString from 'query-string';
import { Icon } from 'UI';

import auth from 'app/Auth';

class ResetPassword extends RouteHandler {
  constructor(props, context) {
    super(props, context);
    this.state = { error: false, password: '', repeatPassword: '' };
    this.submit = this.submit.bind(this);
    this.passwordChange = this.passwordChange.bind(this);
    this.repeatPasswordChange = this.repeatPasswordChange.bind(this);
  }

  passwordChange(e) {
    this.setState({ password: e.target.value });
    this.setState({ passwordError: false });
  }

  repeatPasswordChange(e) {
    this.setState({ repeatPassword: e.target.value });
    this.setState({ passwordError: false });
  }

  submit(e) {
    e.preventDefault();
    const passwordsDontMatch = this.state.password !== this.state.repeatPassword;
    const emptyPassword = this.state.password.trim() === '';
    if (emptyPassword || passwordsDontMatch) {
      this.setState({ error: true });
      return;
    }

    this.props.resetPassword(this.state.password, this.props.params.key).then(() => {
      browserHistory.push('/login');
    });

    this.setState({ password: '', repeatPassword: '' });
  }

  render() {
    return (
      <div className="content login-content">
        <div className="row">
          <div className="col-xs-12 col-sm-4 col-sm-offset-4">
            {queryString.parse(this.context.router.location.search).createAccount === 'true' && (
              <div className="alert alert-info">
                <Icon icon="info-circle" />
                <div>
                  To complete the account creation process, please create a password for your
                  account
                </div>
              </div>
            )}
            <form onSubmit={this.submit}>
              <div className={`form-group login-email ${this.state.error ? 'has-error' : ''}`}>
                <label className="form-group-label" htmlFor="password">
                  Password
                </label>
                <input
                  onChange={this.passwordChange}
                  value={this.state.password}
                  type="password"
                  name="password"
                  id="password"
                  className="form-control"
                />
              </div>
              <div className={`form-group login-password ${this.state.error ? 'has-error' : ''}`}>
                <label className="form-group-label" htmlFor="repeat-password">
                  Repeat Password
                </label>
                <input
                  value={this.state.repeatPassword}
                  onChange={this.repeatPasswordChange}
                  type="password"
                  name="repeat-password"
                  id="repeat-password"
                  className="form-control"
                />
                {this.state.error && <div className="required">Passwords don&rsquo;t match</div>}
              </div>
              <button type="submit" className="btn btn-block btn-lg btn-primary">
                Save password
              </button>
            </form>
          </div>
        </div>
      </div>
    );
  }
}

ResetPassword.propTypes = {
  resetPassword: PropTypes.func,
  params: PropTypes.shape({
    key: PropTypes.string,
  }),
};

function mapDispatchToProps(dispatch) {
  return bindActionCreators({ resetPassword: auth.actions.resetPassword }, dispatch);
}

export default connect(null, mapDispatchToProps)(ResetPassword);
