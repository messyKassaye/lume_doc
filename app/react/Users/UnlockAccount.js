import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import { browserHistory } from 'react-router';
import RouteHandler from 'app/App/RouteHandler';

import auth from 'app/Auth';

export class UnlockAccount extends RouteHandler {
  unlockAccount() {
    const { username, code } = this.props.params;
    this.props
      .unlockAccount({ username, code })
      .then(() => {
        browserHistory.push('/login');
      })
      .catch(() => {
        browserHistory.push('/login');
      });
  }

  componentDidMount() {
    this.unlockAccount();
  }

  render() {
    return (
      <div className="content login-content">
        <div className="row">
          <div className="col-xs-12 col-sm-4 col-sm-offset-4 text-center">Verifying...</div>
        </div>
      </div>
    );
  }
}

UnlockAccount.propTypes = {
  unlockAccount: PropTypes.func,
  params: PropTypes.shape({
    username: PropTypes.string,
    code: PropTypes.string,
  }),
};

function mapDispatchToProps(dispatch) {
  return bindActionCreators(
    {
      unlockAccount: auth.actions.unlockAccount,
    },
    dispatch
  );
}

export default connect(null, mapDispatchToProps)(UnlockAccount);
