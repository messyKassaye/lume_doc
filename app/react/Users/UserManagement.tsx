import React, { useState } from 'react';
import { TabContent, TabLink, Tabs } from 'react-tabs-redux';
import UsersList from 'app/Users/components/UsersList';
import { Translate } from 'app/I18N';
import { UserGroups } from './components/usergroups/UserGroups';

export const UserManagement = () => {
  const [selectedTab, setSelectedTab] = useState('users');
  return (
    <div className="userManagementTabs">
      <Tabs selectedTab={selectedTab} renderActiveTabContentOnly handleSelect={setSelectedTab}>
        <div>
          <ul className="nav">
            <li>
              <TabLink to="users">
                <Translate>Users</Translate>
              </TabLink>
            </li>
            <li>
              <TabLink to="usergroups">
                <Translate>Groups</Translate>
              </TabLink>
            </li>
          </ul>
        </div>
        <TabContent for="users">
          <UsersList />
        </TabContent>
        <TabContent for="usergroups">
          <UserGroups />
        </TabContent>
      </Tabs>
    </div>
  );
};
