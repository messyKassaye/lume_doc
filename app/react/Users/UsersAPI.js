import api from 'app/utils/api';

export default {
  save(requestParams) {
    return api.post('users', requestParams).then(response => response.json);
  },

  new(requestParams) {
    return api.post('users/new', requestParams).then(response => response.json);
  },

  currentUser(requestParams) {
    return api.get('user', requestParams).then(response => response.json);
  },

  get(requestParams) {
    return api.get('users', requestParams).then(response => response.json);
  },

  getById(requestParams) {
    return api.get('users', requestParams).then(response => response.json[0]);
  },

  delete(requestParams) {
    return api.delete('users', requestParams).then(response => response.json);
  },
};
