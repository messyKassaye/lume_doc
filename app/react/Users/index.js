import Users from './Users';
import NewUser from './components/NewUser';
import EditUser from './components/EditUser';

export { Users, NewUser, EditUser };
