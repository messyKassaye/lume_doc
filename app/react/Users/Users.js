import React from 'react';

import RouteHandler from 'app/App/RouteHandler';
import { actions } from 'app/BasicReducer';
import { UserManagement } from 'app/Users/UserManagement';
import UsersAPI from './UsersAPI';

export class Users extends RouteHandler {
  static async requestState(requestParams) {
    const users = await UsersAPI.get(requestParams);
    return [actions.set('users', users)];
  }

  render() {
    return <UserManagement />;
  }
}

export default Users;
