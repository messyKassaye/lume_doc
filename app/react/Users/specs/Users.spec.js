import React from 'react';
import { shallow } from 'enzyme';
import UsersAPI from 'app/Users/UsersAPI';
import { UserManagement } from 'app/Users/UserManagement';
import { Users } from '../Users';

describe('Users', () => {
  let component;
  let context;
  let users;

  beforeEach(() => {
    users = [{ _id: 1, name: 'Batman' }];
    spyOn(UsersAPI, 'get').and.returnValue(Promise.resolve(users));
    context = { store: { getState: () => ({}), dispatch: jasmine.createSpy('dispatch') } };
    component = shallow(<Users />, { context });
  });

  describe('requestState', () => {
    it('should get the current user, and metadata', async () => {
      const request = {};
      const actions = await Users.requestState(request);

      expect(actions).toMatchSnapshot();
      expect(UsersAPI.get).toHaveBeenCalledWith(request);
    });
  });

  describe('render', () => {
    it('should render a UsersList', () => {
      expect(component.find(UserManagement).length).toBe(1);
    });
  });
});
