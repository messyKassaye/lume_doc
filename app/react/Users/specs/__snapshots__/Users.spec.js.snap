// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Users requestState should get the current user, and metadata 1`] = `
Array [
  Object {
    "type": "users/SET",
    "value": Array [
      Object {
        "_id": 1,
        "name": "Batman",
      },
    ],
  },
]
`;
